posts/the-game-of-vue-migrations.md

```diff
@@ -0,0 +1,158 @@
+
+Strongly-typed languages like TypeScript allow for better readability because methods and objects are well-structured. It is all about real-time feedback while developing the working code. Perhaps even more valuable is how the types can prevent exceptions from occurring during runtime too.
+
+### Adding TypeScript and Fixing Inconsistencies
+
+Add [TypeScript](https://cli.vuejs.org/core-plugins/typescript.html) via the Vue CLI by running the command:
+
+```bash
+vue add typescript
+```
+
+**Important note:** The first prompt, "`Use class-style component syntax?`" should be answered "`no`". While it is a popular style for writing [TypeScript components](https://vuejs.org/v2/guide/typescript.html) in Vue 2, it is not fully supported in Vue 3 (for many [reasons](https://github.com/vuejs/rfcs/pull/17#issuecomment-494242121)). Without getting too ahead, answering "`no`" will make the [Options API](https://v3.vuejs.org/guide/typescript-support.html#using-with-options-api) and [Composition API](https://v3.vuejs.org/guide/typescript-support.html#using-with-composition-api) implementations easier later.
+
+Deleting `HelloWorld.vue` and reverting `App.vue`, one of the first observations in `GameOfLife.vue` is how the "`generateEmptyGrid`" method shows a problem when adding the constructed row to the grid. This comes from the incorrect declaration within "`data()`" for the "`grid`" value of "`[[false]]`".
+
+![Incorrect Grid Declaration](https://raw.githubusercontent.com/ippontech/blog-usa/master/images/2021/10/VueIncorrectGridDeclaration.png)
+
+Instead, the initial value for "`grid`" should have been "`[[{id:-1, alive:false}]]`". Changing it resolves the problematic inconsistency. There is no more confusion, thanks to TypeScript.
+
+As is seen requested for "`toggleCell`" above, adding method outputs and inputs is required in TypeScript because [`strict` checking](https://github.com/matthewreed26/game-of-life/blob/unit-jest-typescript/tsconfig.json#L5) is enabled in the `tsconfig.json`. Once done, the app runs as expected! But the tests do not run. Checking `GameOfLife.spec.ts`, it seems the [use of `Vue.extend`](https://github.com/matthewreed26/game-of-life/blob/unit-jest-typescript/src/components/game-of-life/GameOfLife.component.ts#L3) has broken the `GameOfLife` import. Unfortunately, in Vue 2, there is no easy solution other than [this quick fix](https://github.com/vuejs/vue-test-utils/issues/255) on the [`shallowMount` Wrapper type](https://github.com/matthewreed26/game-of-life/blob/unit-jest-typescript/tests/unit/components/game-of-life/GameOfLife.spec.ts#L145). The addition of that means a new tradeoff. For Visual Studio Code users, the IDE's [IntelliSense](https://code.visualstudio.com/docs/editor/intellisense) now cannot help write new tests. In turn, though, it restores functionality to the existing test suite (modifying an "`expect`" will confirm). The good news is an upgrade to Vue 3 will correct this IDE type-checking problem.
```

Comment:
```suggestion
As is seen requested for "`toggleCell`" above, adding method outputs and inputs is required in TypeScript because [`strict` checking](https://github.com/matthewreed26/game-of-life/blob/unit-jest-typescript/tsconfig.json#L5) is enabled in the `tsconfig.json`. Once done, the app runs as expected! **But**, the tests do _NOT_ run. Checking `GameOfLife.spec.ts`, it seems the [use of `Vue.extend`](https://github.com/matthewreed26/game-of-life/blob/unit-jest-typescript/src/components/game-of-life/GameOfLife.component.ts#L3) has broken the `GameOfLife` import. Unfortunately, in Vue 2, there is no easy solution other than [this quick fix](https://github.com/vuejs/vue-test-utils/issues/255) on the [`shallowMount` Wrapper type](https://github.com/matthewreed26/game-of-life/blob/unit-jest-typescript/tests/unit/components/game-of-life/GameOfLife.spec.ts#L145). The addition of that means a new tradeoff. For Visual Studio Code users, the IDEs [IntelliSense](https://code.visualstudio.com/docs/editor/intellisense) cannot help write new tests. In turn, it restores functionality to the existing test suite (modifying an "`expect`" will confirm). The good news is an upgrade to Vue 3 will correct this IDE type-checking problem.
```